docs/architecture/adr-110-remote-mempool.md

```diff
@@ -0,0 +1,364 @@
+
+### High-level architecture
+
+#### Single-process mempool
+
+The simplest possible architecture for a Comet validator with an external
+mempool is as follows.
+
+```mermaid
+flowchart LR
+    comet[Comet Validator]
+    app[Application]
+    mempool[Mempool]
+    user[User]
+
+    comet --> app
+
+    comet -- 4. Reap transactions --> mempool
+    user -- 1. Submit transactions --> app
+    app -- 2. CheckTx --> app
+    app -- 3. Publish transactions --> mempool
+    mempool -- CheckTx --> app
+```
+
+In this flow:
+
+1. Users submit transactions directly to some form of RPC in the application. It
+   is up to operators to secure this RPC endpoint and implement measures to
+   prevent DDoS attacks.
+2. The application validates incoming transactions internally in a similar way
+   to how `CheckTx` currently works.
+3. Transactions that the application deems valid are then sent to the mempool by
+   the application, e.g. through some form of RPC mechanism. The mempool is then
+   expected to propagate these transactions to the rest of the network.
+4. During `PrepareProposal`, a Comet validator will reap transactions from the
+   mempool by way of an RPC call.
+
+Additionally, when transactions are received by the local mempool instance from
+foreign mempool instances, it is expected that the local mempool instance will
+send the transaction to the application for validation (similar to how `CheckTx`
+currently functions).
+
+#### Scalable mempool
+
+The recommended high-level architecture for a modular Comet validator with an
+externalized, scalable mempool, is as follows.
+
+```mermaid
+flowchart LR
+    comet[Comet Validator]
+    app[Application]
+    mempoolrouter[Mempool Router]
+    mempool1[Mempool Instance 1]
+    mempool2[Mempool Instance 2]
+    mempoolN[Mempool Instance N]
+    user[User]
+
+    comet --> app
+
+    comet -- 4. Reap transactions --> mempoolrouter
+    user -- 1. Submit transactions --> app
+    app -- 2. CheckTx --> app
+    app -- 3. Publish transactions --> mempoolrouter
+
+    mempoolrouter --> mempool1
+    mempoolrouter --> mempool2
+    mempoolrouter --> mempoolN
+
+    mempool1 -- CheckTx --> app
+    mempool2 -- CheckTx --> app
+    mempoolN -- CheckTx --> app
+```
+
+Here the interaction is the same as the simple variant discussed earlier, but
+instead of interacting directly with the mempool, all interactions are with some
+form of routing mechanism. For example, this could be a reverse proxy like nginx
+configured to deliver requests to mempool instances in a round-robin fashion.
+
+This configuration would be application-specific, and would need to be set up
+correctly by the operator for the specific application.
+
+### Configuration
+
+The following change to the `config.toml` file is envisaged:
+
+```toml
+[mempool]
+# The type of mempool for this CometBFT node to use.
+#
+# Valid types of mempools supported by CometBFT:
+# - ""       : Default clist mempool with flooding gossip protocol
```

Comment:
I agree. Also, the preferred gossip protocol is independent of whether the mempool type is local or remote. It could have its own separate configuration.